pageserver/src/virtual_file.rs

```diff
@@ -0,0 +1,221 @@
+use std::os::unix::prelude::FileExt;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::RwLock;
+
+// TODO: probe the actual system limit, or at least make this configurable.
+const MAX_OPEN_FILES: usize = 100;
+
+pub struct VirtualFile {
+    /// Lazy handle to the global file descriptor cache. The slot that this points to
+    /// might contain our File, or it may be empty, or it may contain a File that
+    /// belongs to a different VirtualFile.
+    handle: RwLock<SlotHandle>,
+
+    /// File read position
+    pos: u64,
+
+    /// File path
+    path: PathBuf,
+}
+
+struct SlotHandle {
+    /// Index into OPEN_FILES.slots
+    index: usize,
+
+    /// Counter that's incremented every time a different file is stored here.
+    /// To avoid the ABA problem
+    tag: u64,
+}
+
+struct OpenFiles {
+    next: AtomicUsize,
+    slots: Vec<RwLock<Slot>>,
+}
+
+lazy_static! {
+    ///
+    /// This is the global cache of file descriptors (File objects).
+    ///
+    static ref OPEN_FILES: OpenFiles = OpenFiles {
+        next: AtomicUsize::new(0),
+        slots: {
+            // Initialize a vector of MAX_OPEN_FILES slots. The size of the vector
+            // doesn't change after startup. It could make it a plain array, but that's
+            // cumbersome to initialize.
+            let mut slots = Vec::new();
+            for _ in 0..MAX_OPEN_FILES {
+                let slot = Slot {
+                    tag: 0,
+                    recently_used: AtomicBool::new(false),
+                    file: None,
+                };
+                slots.push(RwLock::new(slot));
+            }
+            slots
+        }
+    };
+}
+
+struct Slot {
+    tag: u64,
+    recently_used: AtomicBool,
+    file: Option<File>,
+}
+
+const INVALID_TAG: u64 = u64::MAX;
+
+impl VirtualFile {
+    pub fn new(path: &Path) -> VirtualFile {
+        VirtualFile {
+            handle: RwLock::new(SlotHandle {
+                index: 0,
+                tag: INVALID_TAG,
+            }),
+            pos: 0,
+            path: path.to_path_buf(),
+        }
+    }
+
+    fn with_file<F, R>(&self, mut func: F) -> Result<R, Error>
+    where
+        F: FnMut(&File) -> R,
+    {
+        // Try cached fd first
+        {
+            let handle = self.handle.read().unwrap();
+
+            if handle.tag != INVALID_TAG {
+                let slot = OPEN_FILES.slots[handle.index].read().unwrap();
+                if slot.tag == handle.tag {
+                    if let Some(file) = &slot.file {
+                        slot.recently_used.store(true, Ordering::Relaxed);
+                        return Ok(func(file));
+                    }
+                }
+            }
+        }
+
+        // Upgrade lock and try again
+        let mut handle = self.handle.write().unwrap();
+        if handle.tag != INVALID_TAG {
+            let slot = OPEN_FILES.slots[handle.index].write().unwrap();
```

Comment:
I wonder, is it possible to have a deadlock here?

Reply:
No. This doesn't actually upgrade the lock (there is no function to do that in RwLock anyway), but it releases the old lock and re-acquires it in write mode.

There are two locks that are held simultaneously: the lock in VirtualFile, and the lock in Slot. The lock in VirtualFile is always acquired first.